compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributionPatterns.cpp

```diff
@@ -369,11 +370,155 @@ struct DistributeTransferWriteLayoutAttr final
+      Value packed = packVectorToSupportedWidth(loc, rewriter, result);
+      auto shuffleOp = rewriter.create<gpu::ShuffleOp>(loc, packed, i, size,
+                                                       gpu::ShuffleMode::XOR);
+      Value unpacked =
+          unpackToVector(loc, rewriter, shuffleOp.getShuffleResult(),
+                         result.getType().cast<VectorType>());
+      result = makeArithReduction(rewriter, loc, combiningKind, unpacked,
+                                  result, nullptr, mask);
+    }
+
+    // Reduce packed vector with initial value.
+    Value reducedValue = rewriter.create<vector::ExtractOp>(
+        loc, result, SmallVector<int64_t>{0});
+    for (int i = 1; i < entriesPerVector; i++) {
+      Value next = rewriter.create<vector::ExtractOp>(loc, result,
+                                                      SmallVector<int64_t>{i});
+      reducedValue = makeArithReduction(rewriter, loc, combiningKind,
+                                        reducedValue, next, nullptr, mask);
+    }
+    result = makeArithReduction(rewriter, loc, combiningKind, reducedValue,
+                                mEmpty, nullptr, mask);
+    return result;
+  }
+
+  // This pattern distributes reductions as follows:
+  // First, the data local to a specific thread is reduced.
+  // Then, the data between threads is reduced by emitting appropriate
+  // shuffle instructions.
+  // Currently, only 16 and 32 bit types are supported.
+  // TODO: Add ability to reduce n parallel dims together.
+  LogicalResult matchAndRewrite(vector::MultiDimReductionOp reductionOp,
+                                DistributionSignature &signature,
+                                PatternRewriter &rewriter) const override {
+    auto reductionDims = llvm::to_vector<4>(
+        reductionOp.getReductionDims().getAsRange<IntegerAttr>());
+    // TODO: Add support for reductions along multiple dimensions.
+    if (reductionDims.size() > 1)
+      return failure();
+
+    VectorValue resultVec = dyn_cast<VectorValue>(reductionOp.getResult());
+    // TODO: Support results that are not vectors.
+    if (!resultVec)
+      return failure();
+    LayoutAttr resultLayout = dyn_cast<LayoutAttr>(signature[resultVec]);
+    if (!resultLayout)
+      return failure();
+
+    VectorValue source = reductionOp.getSource();
+    ShapedType sourceType = llvm::cast<ShapedType>(source.getType());
+    // TODO: Add support for (n != 2)-D tensors.
+    if (sourceType.getRank() != 2)
+      return failure();
+
+    LayoutAttr sourceLayout = dyn_cast<LayoutAttr>(signature[source]);
+    if (!sourceLayout)
+      return failure();
+
+    VectorValue acc = dyn_cast<VectorValue>(reductionOp.getAcc());
+    ShapedType accType = llvm::cast<ShapedType>(acc.getType());
+    Type elementType = accType.getElementType();
+    int bitWidth = elementType.getIntOrFloatBitWidth();
+    // TODO: Support additional bitwidths.
+    if ((bitWidth != 16) && (bitWidth != 32))
```

Comment:
The only thing blocking it is lack of testing. I have only tested 16 and 32 bits so just wanted to add some safeguards for the cases that I did not test. But you are right, in theory it should work for other power of 2 bitwidths <= 32.

Reply:
Ok makes sense to me, fine to leave as is.

Main reason I'm asking here is because I find it easiest to have the switch flip for a strategy be in kernel config (where we need to verify things e2e, so that can be opt-in one at a time) and otherwise try to let the patterns succeed on the things they should work with, but can do that later.

Reply:
Sounds good, thanks!